File: csharp/src/Pronesoft.Ecf.Sdk/Client/IntegrationClient.cs
using System;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading;
using System.Threading.Tasks;
using Pronesoft.Ecf.Sdk.Api;
using Pronesoft.Ecf.Sdk.Model;

namespace Pronesoft.Ecf.Sdk.Client
{
    public class IntegrationClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly AuthenticationApi _authApi;
        private readonly IntegrationTokenManager _tokenManager;
        private readonly string _basePath;

        public AssociatedCompaniesApi AssociatedCompanies { get; }
        public AuthenticationApi Authentication { get; }
        public AutomatedCertificationApi AutomatedCertification { get; }
        public CommercialApprovalsApi CommercialApprovals { get; }
        public DigitalCertificatesApi DigitalCertificates { get; }
        public DocumentsReceivedApi DocumentsReceived { get; }
        public DocumentsSentApi DocumentsSent { get; }
        public ECFSubmissionApi EcfSubmission { get; }
        public ReportsApi Reports { get; }
        public TaxSequencesApi TaxSequences { get; }
        public WebhookConfigurationApi Webhooks { get; }

        public IntegrationClient(
            string baseUrl,
            string clientId,
            string clientSecret,
            int refreshSkewSeconds = 300)
        {
            if (string.IsNullOrWhiteSpace(baseUrl))
            {
                throw new ArgumentException("baseUrl is required", nameof(baseUrl));
            }
            if (string.IsNullOrWhiteSpace(clientId))
            {
                throw new ArgumentException("clientId is required", nameof(clientId));
            }
            if (string.IsNullOrWhiteSpace(clientSecret))
            {
                throw new ArgumentException("clientSecret is required", nameof(clientSecret));
            }

            var normalizedBaseUrl = baseUrl.TrimEnd('/');
            _basePath = normalizedBaseUrl;

            var authConfig = new Configuration { BasePath = normalizedBaseUrl };
            _authApi = new AuthenticationApi(authConfig);
            _tokenManager = new IntegrationTokenManager(_authApi, clientId, clientSecret, Math.Max(refreshSkewSeconds, 1));

            var authHandler = new IntegrationAuthHandler(_tokenManager, new HttpClientHandler());
            _httpClient = new HttpClient(authHandler, true);

            var sdkConfig = new Configuration { BasePath = normalizedBaseUrl, AccessToken = "integration-managed" };
            AssociatedCompanies = new AssociatedCompaniesApi(_httpClient, sdkConfig);
            Authentication = new AuthenticationApi(_httpClient, sdkConfig);
            AutomatedCertification = new AutomatedCertificationApi(_httpClient, sdkConfig);
            CommercialApprovals = new CommercialApprovalsApi(_httpClient, sdkConfig);
            DigitalCertificates = new DigitalCertificatesApi(_httpClient, sdkConfig);
            DocumentsReceived = new DocumentsReceivedApi(_httpClient, sdkConfig);
            DocumentsSent = new DocumentsSentApi(_httpClient, sdkConfig);
            EcfSubmission = new ECFSubmissionApi(_httpClient, sdkConfig);
            Reports = new ReportsApi(_httpClient, sdkConfig);
            TaxSequences = new TaxSequencesApi(_httpClient, sdkConfig);
            Webhooks = new WebhookConfigurationApi(_httpClient, sdkConfig);
        }

        public void Dispose()
        {
            _httpClient.Dispose();
            _authApi.Dispose();
        }

        public TenantScopedClient ForTenant(string tenantId)
        {
            if (string.IsNullOrWhiteSpace(tenantId))
            {
                throw new ArgumentException("tenantId is required", nameof(tenantId));
            }

            var cfg = new Configuration
            {
                BasePath = _basePath,
                AccessToken = "integration-managed",
                DefaultHeaders = { ["x-tenant-id"] = tenantId.Trim() }
            };
            return new TenantScopedClient(_httpClient, cfg);
        }

        public TenantSessionClient StartSession(string initialTenantId = null)
        {
            return new TenantSessionClient(this, initialTenantId);
        }

        public sealed class TenantScopedClient
        {
            public AssociatedCompaniesApi AssociatedCompanies { get; }
            public AuthenticationApi Authentication { get; }
            public AutomatedCertificationApi AutomatedCertification { get; }
            public CommercialApprovalsApi CommercialApprovals { get; }
            public DigitalCertificatesApi DigitalCertificates { get; }
            public DocumentsReceivedApi DocumentsReceived { get; }
            public DocumentsSentApi DocumentsSent { get; }
            public ECFSubmissionApi EcfSubmission { get; }
            public ReportsApi Reports { get; }
            public TaxSequencesApi TaxSequences { get; }
            public WebhookConfigurationApi Webhooks { get; }

            internal TenantScopedClient(HttpClient httpClient, Configuration config)
            {
                AssociatedCompanies = new AssociatedCompaniesApi(httpClient, config);
                Authentication = new AuthenticationApi(httpClient, config);
                AutomatedCertification = new AutomatedCertificationApi(httpClient, config);
                CommercialApprovals = new CommercialApprovalsApi(httpClient, config);
                DigitalCertificates = new DigitalCertificatesApi(httpClient, config);
                DocumentsReceived = new DocumentsReceivedApi(httpClient, config);
                DocumentsSent = new DocumentsSentApi(httpClient, config);
                EcfSubmission = new ECFSubmissionApi(httpClient, config);
                Reports = new ReportsApi(httpClient, config);
                TaxSequences = new TaxSequencesApi(httpClient, config);
                Webhooks = new WebhookConfigurationApi(httpClient, config);
            }
        }

        public sealed class TenantSessionClient
        {
            private readonly IntegrationClient _integrationClient;
            private readonly object _sync = new object();
            private string _currentTenantId;

            internal TenantSessionClient(IntegrationClient integrationClient, string initialTenantId = null)
            {
                _integrationClient = integrationClient;
                if (!string.IsNullOrWhiteSpace(initialTenantId))
                {
                    SetCurrentTenant(initialTenantId);
                }
            }

            public TenantSessionClient SetCurrentTenant(string tenantId)
            {
                if (string.IsNullOrWhiteSpace(tenantId))
                {
                    throw new ArgumentException("tenantId is required", nameof(tenantId));
                }

                lock (_sync)
                {
                    _currentTenantId = tenantId.Trim();
                }

                return this;
            }

            public TenantSessionClient ClearCurrentTenant()
            {
                lock (_sync)
                {
                    _currentTenantId = null;
                }

                return this;
            }

            public string GetCurrentTenantId()
            {
                lock (_sync)
                {
                    return _currentTenantId;
                }
            }

            public TenantScopedClient GetClient()
            {
                var tenantId = GetCurrentTenantId();
                if (string.IsNullOrWhiteSpace(tenantId))
                {
                    throw new InvalidOperationException("Current tenant is not set");
                }

                return _integrationClient.ForTenant(tenantId);
            }
        }

        private sealed class IntegrationTokenManager
        {
            private readonly AuthenticationApi _authApi;
            private readonly string _clientId;
            private readonly string _clientSecret;
            private readonly int _refreshSkewSeconds;
            private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);

            private string _accessToken;
            private DateTimeOffset _expiresAt;

            public IntegrationTokenManager(AuthenticationApi authApi, string clientId, string clientSecret, int refreshSkewSeconds)
            {
                _authApi = authApi;
                _clientId = clientId;
                _clientSecret = clientSecret;
                _refreshSkewSeconds = refreshSkewSeconds;
            }

            public async Task<string> GetTokenAsync(bool forceRefresh, CancellationToken cancellationToken)
            {
                var now = DateTimeOffset.UtcNow;
                if (!forceRefresh && !string.IsNullOrWhiteSpace(_accessToken) && now < _expiresAt.AddSeconds(-_refreshSkewSeconds))
                {
                    return _accessToken;
                }

                await _lock.WaitAsync(cancellationToken).ConfigureAwait(false);
                try
                {
                    now = DateTimeOffset.UtcNow;
                    if (!forceRefresh && !string.IsNullOrWhiteSpace(_accessToken) && now < _expiresAt.AddSeconds(-_refreshSkewSeconds))
                    {
                        return _accessToken;
                    }

                    var token = await _authApi.GetAccessTokenAsync(new OAuthTokenRequest(_clientId, _clientSecret), cancellationToken)
                        .ConfigureAwait(false);

                    if (string.IsNullOrWhiteSpace(token.AccessToken))
                    {
                        throw new InvalidOperationException("OAuth token response did not include accessToken");
                    }

                    var expiresIn = token.ExpiresIn > 0 ? token.ExpiresIn : 86400;
                    _accessToken = token.AccessToken;
                    _expiresAt = DateTimeOffset.UtcNow.AddSeconds(expiresIn);
                    return _accessToken;
                }
                finally
                {
                    _lock.Release();
                }
            }
        }

        private sealed class IntegrationAuthHandler : DelegatingHandler
        {
            private readonly IntegrationTokenManager _tokenManager;

            public IntegrationAuthHandler(IntegrationTokenManager tokenManager, HttpMessageHandler innerHandler)
                : base(innerHandler)
            {
                _tokenManager = tokenManager;
            }

            protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            {
                if (request.RequestUri != null && request.RequestUri.AbsolutePath.EndsWith("/oauth/token", StringComparison.Ordinal))
                {
                    return await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
                }

                var token = await _tokenManager.GetTokenAsync(false, cancellationToken).ConfigureAwait(false);
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

                var response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
                if (response.StatusCode != HttpStatusCode.Unauthorized)
                {
                    return response;
                }

                response.Dispose();

                var refreshedToken = await _tokenManager.GetTokenAsync(true, cancellationToken).ConfigureAwait(false);
                var retryRequest = await CloneRequestAsync(request).ConfigureAwait(false);
                retryRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", refreshedToken);
                return await base.SendAsync(retryRequest, cancellationToken).ConfigureAwait(false);
            }

            private static async Task<HttpRequestMessage> CloneRequestAsync(HttpRequestMessage request)
            {
                var clone = new HttpRequestMessage(request.Method, request.RequestUri)
                {
                    Version = request.Version
                };

                foreach (var header in request.Headers)
                {
                    clone.Headers.TryAddWithoutValidation(header.Key, header.Value);
                }

                if (request.Content != null)
                {
                    var ms = new MemoryStream();
                    await request.Content.CopyToAsync(ms).ConfigureAwait(false);
                    ms.Position = 0;
                    var streamContent = new StreamContent(ms);
                    foreach (var header in request.Content.Headers)
                    {
                        streamContent.Headers.TryAddWithoutValidation(header.Key, header.Value);
                    }
                    clone.Content = streamContent;
                }

                return clone;
            }
        }
    }
}

File: java/src/main/java/com/pronesoft/ecf/IntegrationClient.java
package com.pronesoft.ecf;

import com.pronesoft.ecf.api.AssociatedCompaniesApi;
import com.pronesoft.ecf.api.AuthenticationApi;
import com.pronesoft.ecf.api.AutomatedCertificationApi;
import com.pronesoft.ecf.api.CommercialApprovalsApi;
import com.pronesoft.ecf.api.DigitalCertificatesApi;
import com.pronesoft.ecf.api.DocumentsReceivedApi;
import com.pronesoft.ecf.api.DocumentsSentApi;
import com.pronesoft.ecf.api.ECFSubmissionApi;
import com.pronesoft.ecf.api.ReportsApi;
import com.pronesoft.ecf.api.TaxSequencesApi;
import com.pronesoft.ecf.api.WebhookConfigurationApi;

public class IntegrationClient {
    private final ApiClient apiClient;
    private final String normalizedBaseUrl;
    private final String clientId;
    private final String clientSecret;

    public final AssociatedCompaniesApi associatedCompanies;
    public final AuthenticationApi authentication;
    public final AutomatedCertificationApi automatedCertification;
    public final CommercialApprovalsApi commercialApprovals;
    public final DigitalCertificatesApi digitalCertificates;
    public final DocumentsReceivedApi documentsReceived;
    public final DocumentsSentApi documentsSent;
    public final ECFSubmissionApi ecfSubmission;
    public final ReportsApi reports;
    public final TaxSequencesApi taxSequences;
    public final WebhookConfigurationApi webhooks;

    public IntegrationClient(String baseUrl, String clientId, String clientSecret) {
        if (baseUrl == null || baseUrl.trim().isEmpty()) {
            throw new IllegalArgumentException("baseUrl is required");
        }
        if (clientId == null || clientId.trim().isEmpty()) {
            throw new IllegalArgumentException("clientId is required");
        }
        if (clientSecret == null || clientSecret.trim().isEmpty()) {
            throw new IllegalArgumentException("clientSecret is required");
        }

        String normalizedBaseUrl = baseUrl.replaceAll("/+$", "");

        this.normalizedBaseUrl = normalizedBaseUrl;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.apiClient = new ApiClient(normalizedBaseUrl, clientId, clientSecret, null);

        this.associatedCompanies = new AssociatedCompaniesApi(this.apiClient);
        this.authentication = new AuthenticationApi(this.apiClient);
        this.automatedCertification = new AutomatedCertificationApi(this.apiClient);
        this.commercialApprovals = new CommercialApprovalsApi(this.apiClient);
        this.digitalCertificates = new DigitalCertificatesApi(this.apiClient);
        this.documentsReceived = new DocumentsReceivedApi(this.apiClient);
        this.documentsSent = new DocumentsSentApi(this.apiClient);
        this.ecfSubmission = new ECFSubmissionApi(this.apiClient);
        this.reports = new ReportsApi(this.apiClient);
        this.taxSequences = new TaxSequencesApi(this.apiClient);
        this.webhooks = new WebhookConfigurationApi(this.apiClient);
    }

    public ApiClient getApiClient() {
        return this.apiClient;
    }

    public TenantScopedClient forTenant(String tenantId) {
        if (tenantId == null || tenantId.trim().isEmpty()) {
            throw new IllegalArgumentException("tenantId is required");
        }

        ApiClient tenantApiClient = new ApiClient(this.normalizedBaseUrl, this.clientId, this.clientSecret, null);
        tenantApiClient.addDefaultHeader("x-tenant-id", tenantId.trim());
        return new TenantScopedClient(tenantApiClient);
    }

    public TenantSessionClient startSession(String initialTenantId) {
        return new TenantSessionClient(this, initialTenantId);
    }

    public static class TenantScopedClient {
        public final AssociatedCompaniesApi associatedCompanies;
        public final AuthenticationApi authentication;
        public final AutomatedCertificationApi automatedCertification;
        public final CommercialApprovalsApi commercialApprovals;
        public final DigitalCertificatesApi digitalCertificates;
        public final DocumentsReceivedApi documentsReceived;
        public final DocumentsSentApi documentsSent;
        public final ECFSubmissionApi ecfSubmission;
        public final ReportsApi reports;
        public final TaxSequencesApi taxSequences;
        public final WebhookConfigurationApi webhooks;

        TenantScopedClient(ApiClient apiClient) {
            this.associatedCompanies = new AssociatedCompaniesApi(apiClient);
            this.authentication = new AuthenticationApi(apiClient);
            this.automatedCertification = new AutomatedCertificationApi(apiClient);
            this.commercialApprovals = new CommercialApprovalsApi(apiClient);
            this.digitalCertificates = new DigitalCertificatesApi(apiClient);
            this.documentsReceived = new DocumentsReceivedApi(apiClient);
            this.documentsSent = new DocumentsSentApi(apiClient);
            this.ecfSubmission = new ECFSubmissionApi(apiClient);
            this.reports = new ReportsApi(apiClient);
            this.taxSequences = new TaxSequencesApi(apiClient);
            this.webhooks = new WebhookConfigurationApi(apiClient);
        }
    }

    public static class TenantSessionClient {
        private final IntegrationClient integrationClient;
        private String currentTenantId;

        TenantSessionClient(IntegrationClient integrationClient, String initialTenantId) {
            this.integrationClient = integrationClient;
            if (initialTenantId != null && !initialTenantId.trim().isEmpty()) {
                setCurrentTenant(initialTenantId);
            }
        }

        public TenantSessionClient setCurrentTenant(String tenantId) {
            if (tenantId == null || tenantId.trim().isEmpty()) {
                throw new IllegalArgumentException("tenantId is required");
            }
            this.currentTenantId = tenantId.trim();
            return this;
        }

        public TenantSessionClient clearCurrentTenant() {
            this.currentTenantId = null;
            return this;
        }

        public String getCurrentTenantId() {
            return this.currentTenantId;
        }

        public TenantScopedClient getClient() {
            if (this.currentTenantId == null || this.currentTenantId.isEmpty()) {
                throw new IllegalStateException("Current tenant is not set");
            }
            return this.integrationClient.forTenant(this.currentTenantId);
        }
    }
}
